pkg/provision/ec2.go

```diff
@@ -0,0 +1,203 @@
+package provision
+
+import (
+	"fmt"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"strconv"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/google/uuid"
+)
+
+// EC2Provisioner contains the EC2 client
+type EC2Provisioner struct {
+	ec2Provisioner *ec2.EC2
+}
+
+// NewEC2Provioner creates an EC2Provisioner and initialises an EC2 client
+func NewEC2Provisioner(region, accessKey, secretKey string) (*EC2Provisioner, error) {
+	sess, err := session.NewSession(&aws.Config{
+		Region:      aws.String(region),
+		Credentials: credentials.NewStaticCredentials(accessKey, secretKey, ""),
+	})
+	svc := ec2.New(sess)
+	return &EC2Provisioner{ec2Provisioner: svc}, err
+}
+
+// Provision deploys an exit node into AWS
+func (p *EC2Provisioner) Provision(host BasicHost) (*ProvisionedHost, error) {
+	image, err := p.lookupAMI(host.OS)
+	if err != nil {
+		return nil, err
+	}
+
+	groupID, name, err := p.securityGroup(host.Additional["inlets-port"])
+	if err != nil {
+		return nil, err
+	}
+
+	runResult, err := p.ec2Provisioner.RunInstances(&ec2.RunInstancesInput{
+		ImageId:      image,
+		InstanceType: aws.String(host.Plan),
+		MinCount:     aws.Int64(1),
+		MaxCount:     aws.Int64(1),
+		UserData:     &host.UserData,
+		NetworkInterfaces: []*ec2.InstanceNetworkInterfaceSpecification{
+			{
+				DeviceIndex:              aws.Int64(int64(0)),
+				AssociatePublicIpAddress: aws.Bool(true),
+				DeleteOnTermination:      aws.Bool(true),
+				Groups:                   []*string{groupID},
+			},
+		},
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	err = p.ec2Provisioner.WaitUntilInstanceRunning(&ec2.DescribeInstancesInput{
```

Comment:
I remember a conversation where we discussed blocking vs. non-blocking and agreed to remove the blocking call. I'm seeing the word "WaitUntil" here, so wanted to check in?

Reply:
@alexellis, this blocking only blocks until the instance is running which is a few seconds, if this is not in place then the first status check will fail with an instance not found. The previous blocking blocked until the node was fully functional.

Reply:
I just read your other comment, if we remove the lookups from the status method then we do not need this wait. I will update.

Reply:
I have removed this.